hbase-balancer/src/main/java/org/apache/hadoop/hbase/master/balancer/BalancerClusterState.java

```diff
@@ -207,15 +212,20 @@ public String getRack(ServerName server) {
 
     serverIndexToHostIndex = new int[numServers];
     serverIndexToRackIndex = new int[numServers];
-    regionsPerServer = new int[numServers][];
-    serverIndexToRegionsOffset = new int[numServers];
-    regionsPerHost = new int[numHosts][];
-    regionsPerRack = new int[numRacks][];
-    primariesOfRegionsPerServer = new int[numServers][];
-    primariesOfRegionsPerHost = new int[numHosts][];
-    primariesOfRegionsPerRack = new int[numRacks][];
+    regionsPerServer = new ArrayList<ArrayList<Integer>>(numServers);
```

Comment:
I think we could implement a simple IntArrayList, this could save a lot of memory, so here we could write code like

regionsPerServer = new IntArrayList[numServers];

Reply:
This is a two dimensional collection. The same as regionsPerHost and regionPerRack which are changed from int[][] to ArrayList<HashSet<Integer>> for lookup/insert and delete by value. There are methods that take all three as input. The best I can do is to make all three as array of collections(HashSet or int[]). The biggest concerns I have is to mix generic and primitives and arrays which is generally not recommended. Let me test the perf difference and report back if it is worth it. Thank you.